### PR TITLE
NEW: GraphQL 4 Compatibility

### PR DESCRIPTION
**Pull request by unclecheese**
Barely working. Blocks render, can add a new element.

### CODE REVIEW

**Review by chillu — CHANGES_REQUESTED** [review]

File: src/GraphQL/Resolvers/Resolver.php [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+            $reorderer->reorder($elementID);
+
+            return $clone;
+        } catch (Exception $e) {
+            throw new Exception("Something went wrong when duplicating element: $elementID");
+        }
+    }
+
+
+    public static function newTitle(string $title): string
```
Comment (chillu) [inline src/GraphQL/Resolvers/Resolver.php:157]: Title is optional (can be blank in new blocks) - regression from GraphQL v3 functionality.

File: client/src/state/history/readOneBlockQuery.js [hunk 2]
```diff
@@ -7,36 +7,31 @@ import gql from 'graphql-tag';
 const query = gql`
 query ReadHistoryViewerBlock ($block_id: ID!, $limit: Int!, $offset: Int!) {
   readOneBlock(
-    Versioning: {
-      Mode: LATEST
+    versioning: {
+      mode: LATEST
     },
-    ID: $block_id
+    id: $block_id
```
Comment (chillu) [inline client/src/state/history/readOneBlockQuery.js:13]: This complains when using with GraphQL v4:

```
query ReadHistoryViewerBlock($block_id: ID!, $limit: Int!, $offset: Int!) {↵  readOneBlock(versioning: {mode: LATEST}, id: $block_id) {...
```

```
{"data":null,"errors":[{"message":"Unknown argument \"id\" on field \"readOneBlock\" of type \"Query\".","code":0,
```

File: _graphql/scalars.yml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+ObjectType:
```
[the remark this answers is not in the document]
Comment (unclecheese) [reply _graphql/scalars.yml:1]: blockSchema field on baseElement is just a JSON blob.

[the remark this answers is not in the document]
Comment (unclecheese) [reply _graphql/scalars.yml:1]: This should actually be core.. I wouldn't mind having a JSON type added as a custom scalar for all schemas. It's useful.

File: src/GraphQL/Resolvers/Resolver.php (hunk 1 above)
Comment (unclecheese) [reply src/GraphQL/Resolvers/Resolver.php:157]: Are you saying the title should be nullable? I'm looking at this: https://github.com/dnadesign/silverstripe-elemental/blob/4/src/GraphQL/DuplicateElementMutation.php#L78

File: client/src/state/history/readOneBlockQuery.js (hunk 2 above)
Comment (unclecheese) [reply client/src/state/history/readOneBlockQuery.js:13]: Have resolved this.

File: _graphql/scalars.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (unclecheese) [reply _graphql/scalars.yml:1]: Tracking https://github.com/silverstripe/silverstripe-graphql/issues/321

File: src/GraphQL/Resolvers/Resolver.php (hunk 1 above)
Comment (chillu) [reply src/GraphQL/Resolvers/Resolver.php:157]: Needed null coalescing, see bfa278044ef9ef9fe13ef1cc82c2c7e1bc1ad95e

**Review by chillu — APPROVED** [review]

File: client/src/state/history/readOneBlockQuery.js [hunk 4]
```diff
@@ -7,36 +7,31 @@ import gql from 'graphql-tag';
...
-    ID: $block_id
+    filter: { id: { eq: $block_id } }
   ) {
-    ID
-    Versions (limit: $limit, offset: $offset, sortBy: {
-      field: Version
-      direction: DESC
-    }) {
+    id
+    versions (limit: $limit, offset: $offset, sort: { version: DESC }) {
```
Comment (chillu) [inline client/src/state/history/readOneBlockQuery.js:16]: ReadHistoryViewerBlock query fails with
{"data":null,"errors":[{"message":"Unknown argument \"limit\" on field \"versions\" of type \"Block\"."

**Comment by unclecheese** [comment]
Moved to https://github.com/dnadesign/silverstripe-elemental/pull/847

**Pull request closed without merging**